src/main/java/org/commcare/formplayer/services/QueryRequester.java

```diff
@@ -2,38 +2,72 @@
 import org.springframework.http.HttpEntity;
 import org.springframework.http.HttpHeaders;
 import org.springframework.http.HttpMethod;
 import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Component;
 import org.springframework.web.client.RestTemplate;
 
 import java.io.UnsupportedEncodingException;
 import java.net.URLDecoder;
+import java.util.concurrent.TimeUnit;
 
+import javax.annotation.Resource;
 
+
+@Component
 public class QueryRequester {
 
+    @Autowired
+    private RedisTemplate redisTemplateString;
+
+    @Autowired
+    private RestoreFactory restoreFactory;
+
+    @Resource(name = "redisTemplateString")
+    private ValueOperations<String, String> caseSearchCache;
+
     private final Log log = LogFactory.getLog(QueryRequester.class);
 
     public String makeQueryRequest(String uri, HttpHeaders headers) {
-        RestTemplate restTemplate = new RestTemplate();
-        ResponseEntity<String> response =
-                null;
-        try {
-            response = restTemplate.exchange(
-                    // Spring framework automatically encodes urls. This ensures we don't pass in an already
-                    // encoded url.
-                    URLDecoder.decode(uri, "UTF-8"),
-                    HttpMethod.GET,
-                    new HttpEntity<String>(headers),
-                    String.class
-            );
-        } catch (UnsupportedEncodingException e) {
-            throw new RuntimeException(e);
+        String responseBody = loadResponseFromCache(uri);
+        if (responseBody == null) {
+            RestTemplate restTemplate = new RestTemplate();
+            ResponseEntity<String> response;
+            try {
+                response = restTemplate.exchange(
+                        // Spring framework automatically encodes urls. This ensures we don't pass in an already
+                        // encoded url.
+                        URLDecoder.decode(uri, "UTF-8"),
+                        HttpMethod.GET,
+                        new HttpEntity<String>(headers),
+                        String.class
+                );
+            } catch (UnsupportedEncodingException e) {
+                throw new RuntimeException(e);
+            }
+            responseBody = response.getBody();
+            if (response.getStatusCode().is2xxSuccessful()) {
+                cacheResponse(uri, responseBody);
+            }
+            log.info(String.format("Query request to URL %s successful", uri));
         }
-        String responseBody = response.getBody();
-        log.info(String.format("Query request to URL %s successful", uri));
         return responseBody;
     }
+
+    private String loadResponseFromCache(String uri) {
+        return caseSearchCache.get(getCacheKey(uri));
+    }
+
+    private void cacheResponse(String uri, String responseBody) {
+        caseSearchCache.set(getCacheKey(uri), responseBody, 2, TimeUnit.MINUTES);
```

Comment:
I might start with 5 minutes just because it probably takes a bit to make it through the menus all the way to the form session

Reply:
would it be ok to not show user a new case in search as long as 5 mins ? I am not sure how realtime we generally expect  our systems to perform in this regard.

Reply:
I would not expect any users to be refreshing he same view looking for new results. I think they generally go through the search, find the case they need, then a number of subsequent requests repeat the search because of formplayer structure. Honestly, I think it would be more confusing for a given user session if the results changed, than if they didn't. And five minutes is short enough that by the time they complete their work (fill out the form), it will have long since expired, so the next query will be fresh.

Reply:
makes sense. changed to 5 mins [here](https://github.com/dimagi/formplayer/pull/850/commits/8e7c83991316e61f2e7eb74caf72288bfb0e19a6)